### PR TITLE
Add alternate (betaflight) rates to copter in Acro mode

### PR DESCRIPTION
**Pull request by chronowarper**
Added support for alternative rates. Betaflight rates are added in this commit.

Roll/pitch rates will apply in ACRO mode.
Yaw rates will apply in all modes

Rate parameters are under ARATE_*
ARATE_TYPE = 0 to use default ardupilot rates (any value that's not defined will also default to ardupilot rates)
ARATE_TYPE = 1 to use betaflight rates
ARATE_BF_RP* and ARATE_BF_Y_* = set the betaflight roll/pitch and yaw RC,Super,Expo values.

Tested RPY on the bench with props off with two sets of BF values by comparing RATE_DES against RC_IN in the logs.
Also passed build_all.sh

### CODE REVIEW

**Comment by peterbarker** [comment]
There are several stylistic things that would need to be fixed before merge.

Marking for DevCall to make sure we'd accept such a patch before bothering you with those.

@chronowarper you covered a lot of "what" in your PR description here but not a lot of the "why" :-)

**Comment by chronowarper** [comment]
Thanks for reviewing. 'Why' is a good question :)  

Main reasons for me: 
  BF rates allows for semi-independent adjustment of rates at the >60% deflection and around center stick. The ardupilot default formula doesn't do this very well with just one expo value and I couldn't get what I wanted (low-sloped linear response around middle with a really high end point)

  I have a large number of mini-quads running betaflight and I wanted to have my ardupilot mini-quad to have the same rates so muscle memory carried over.


I also just wanted to see if I can code this and learn a bit more about ardupilot. I'd be excited to fix it up if it gets accepted. Otherwise it's just for myself and anyone who's interested.

**Comment by andyp1per** [comment]
I doubt we would allow the addition of 7 parameters simply to get the pilot response right. Is there any way of reducing these to the ones that matter. Would it for instance be possible to have some presets that can be modified by 1 or 2 parameters?

**Comment by chronowarper** [comment]
The BF rate curves are very programmable and rates tend to be very personal, so I'm not sure how well the presets will work for everyone. 

I thought of these options to reduce parameter count:

I can reduce the enable parameter by adding a flag in ACRO_OPTIONS instead... I actually started with this but the bitmask didn't show up in mission planner so I'd need some help to get this working.

For the rates, I could:
#1. overload the use of the existing parameters (ie. use ACRO_*_P for BF_RC, and ACRO_*_EXPO for BF_EXP and add 2 parameters for super rate. The risk is the rates will be out of whack if user went back to default option and forgot to update the BF rates.

or

#2. use a uint32 and encode the BF rates for RP and YAW:
  uint32 BF_RP  [7:0] = RC,     //0 ~ 255
                        [14:8] = Super,   // 0 ~ 99
                        [22:16] = Expo   // 0 ~ 100  

 can drop this down to a UINT16 by making step larger. 

Another way to look at #2 is it's effectively a huge, sparse preset list.

**Comment by andyp1per** [comment]
Is it possible to graph the response you are looking for relative to what you can achieve with the current parameters? I can't quite get my head around where you are trying to get to and I suspect @lthall could cookup an expo function that gives you what you want quite easily - for instance by adding second or third order terms

**Comment by IamPete1** [comment]
As said above I think we can get quite close with the existing params. We have a long standing issue for converting the Roll/pitch/Yaw P terms to actual rates. If we were to do that first this change set would be only two second order curve params. In that case we might be able to just use them all the time with the default of 0 so as not to change the current behavior. 

https://github.com/ArduPilot/ardupilot/issues/1521

**Comment by chronowarper** [comment]
here's an example: 

image
image

With Actual rates (playing with this calculator) https://www.desmos.com/calculator/r5pkxlxhtb:

I can get pretty close to the BF rates, so that's a viable alternative. It seems the extra param would be the Actual_center-rate (so default 0 is safe). The existing rates would map to the Actual-max_rate and Actual_expo.

**Comment by chronowarper** [comment]
So clarification of terminology, what I refer to as "Actual Rate" is the formula used in the betaflight configurator as "acutal rate". 

The "Actual rate" equation's goes to the 5th power which allows it to push the slope region much further on stick travel than the existing ardupilot equation. Then a lower order formula gets added on to adjust the center stick travel. 

So fundamentally it seems hard not to change the existing ardupilot expo behavior to get the equivalent of BF or "Actual" rates. But definitely possible to do this without adding more than 2 PARAMS.

I'll attach my spreadsheet:

[Rates_tuning_betaflight_arducopter(external)(.xlsx](https://github.com/ArduPilot/ardupilot/files/7165960/Rates_tuning_betaflight_arducopter.external.xlsx)

File: ArduCopter/mode_acro.cpp [hunk 1]
```diff
@@ -110,11 +110,40 @@ void ModeAcro::get_pilot_desired_angle_rates(int16_t roll_in, int16_t pitch_in,
...
+        // roll axis
+        rp_in = abs(float(roll_in))/ROLL_PITCH_YAW_INPUT_MAX;   //calculate input ratio. Need to take absolute value as input is signed
+        rp_sign= roll_in/abs(roll_in);                          //get sign for final rate calculation            
+
+        bf_p = 1.0f/(1.0f-(rp_in*g.altrate_bf_rp_super));
+        bf_q = (rp_in*rp_in*rp_in*rp_in*g.altrate_bf_rp_expo)+rp_in*(1.0f-g.altrate_bf_rp_expo);
+        rate_bf_request.x = rp_sign*20000.0f*bf_q*g.altrate_bf_rp_rc*bf_p;      //ardupilot rate is in units of centidegrees
+        
+        // pitch axis
+        rp_in = abs(float(pitch_in))/ROLL_PITCH_YAW_INPUT_MAX;    //calculate input ratio. Need to take absolute value as input is signed
```
Comment (tridge) [inline ArduCopter/mode_acro.cpp:138]: should probably be fabsf() not abs(), or you get an integer

File: ArduCopter/mode_acro.cpp [hunk 2]
```diff
@@ -110,11 +110,40 @@ void ModeAcro::get_pilot_desired_angle_rates(int16_t roll_in, int16_t pitch_in,
...
+        rp_sign= pitch_in/abs(pitch_in);                          //get sign for final rate alculation
```
Comment (tridge) [inline ArduCopter/mode_acro.cpp:139]: fabsf() here too?
or use signbitf() ?

File: ArduCopter/mode_acro.cpp [hunk 3]
```diff
@@ -110,11 +110,40 @@ void ModeAcro::get_pilot_desired_angle_rates(int16_t roll_in, int16_t pitch_in,
...
+        float rp_in,rp_sign,bf_p,bf_q;
+        
+        // range check BF rates
+        g.altrate_bf_rp_rc = constrain_float(g.altrate_bf_rp_rc, 0.0f, 2.55f);
+        g.altrate_bf_rp_super = constrain_float(g.altrate_bf_rp_super, 0.0f, 0.99f);    
+        g.altrate_bf_rp_expo = constrain_float(g.altrate_bf_rp_expo, 0.0f, 1.0f);
+
+                
+        // roll axis
+        rp_in = abs(float(roll_in))/ROLL_PITCH_YAW_INPUT_MAX;   //calculate input ratio. Need to take absolute value as input is signed
```
Comment (tridge) [inline ArduCopter/mode_acro.cpp:130]: fabsf()

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
really needs some input from Leonard

**Comment by lthall** [comment]
First up nice work and thanks for taking the time to get this working. It is good to start getting input into these variables as they have not been heavily used.

There are a couple key things I noticed looking at the results of your graphs that I think we can learn from BetaFlight after looking at the various parts of these equations.
image
1. The x^4 function provides a more continuous curve but with a more linier centre section than the x^3 function.
2. The x^4 function can provide up to a 4:1 slope change while the x^3 can only achieve 3:1
3. The "super rate" is achieved by dividing by (1-SR*input) to achieve a sharp increase at the high end.

I note this quote in the BetaFlight description "Although Super Rate has replaced the primary function of Expo, Expo can still be used in conjunction with RC Rate and Super Rate if an extremely dull centre stick region is desired."

So the key to the "super rate" is that it is better than the normal Expo calculation (both x^3 and x^4) at achieving reasonable zero stick response and high maximum stick response. Here is an example of all three set such that the maximum stick slope is the same:
image
image

Given what I see here I would throw away our expo and just use the formulation of super rate I show here where the variable defines the slope of the line about zero as a proportion of maximum rate.

**Comment by chronowarper** [comment]
Thanks Leonard for the analysis and feedback. the issue with x^3 was definitely part of this. I'm clear if that meant we'd prefer to keep the rates with 2 parameters, so I wanted to give my perspective on why I believe 3 parameters is beneficial:

After playing with the different BF rate types I realized that I always wanted to adjust these three things
1. slope near center stick
2. slope near end stick
3. transition point (where it transition from a slow 'center' response to fast 'end' response). 

These are 3 independent degrees of freedom, so they can only be covered by 3 parameters. So while replacing the expo formula with a different one may change #2 and #1, we still don't get the ability to adjust #3. 

image

IMO #3 is an important adjustment for stick-feel as we all have different comfort levels (due to radio size and adjustment, hand size, thumb vs pinch) in how far we can deflect the stick with 'fine' control. This year I've significantly improved the precision of my flying by taking a closer look at what my 'natural' reach is and then mapping the that to how I want the rates to be. 


If we agree on this then we're just left with picking the formula.

**Comment by andyp1per** [comment]
I think the super expo formula might actually work here, it changes the transition point as your diagram shows. 1/2/3 are all related rather than being completely independent. It would be worth seeing how well the super rate function maps to your desired curve.

**Comment by IamPete1** [comment]
I wonder if we should take the opportunity to move the rate and expo params down into attitude control so we can share between copter and quadplane.

**Comment by chronowarper** [comment]
Changing EXPO will naturally change the transition point, but that doesn't decouple the slope from where the transition point occurs. Yes the 3 behaviors are tied together but to pseudo-independently adjust the 3 would require 3 axis of control and not 2. Mathematically it's impossible to do this with 2.

**Comment by andyp1per** [comment]
I understand that 3 parameters will give maximum flexibility but it means the acro expo configuration increases from 4 to 10 parameters overall which creates a maintenance and user headache - so we are looking for the sweet spot of maximum useful flexibility and minimum configuration.

**Comment by lthall** [comment]
> These are 3 independent degrees of freedom, so they can only be covered by 3 parameters.

I agree with this statement but what I don't agree that these parameters provide this flexibility. And I don't think a pilot could tell between any combination of even the two Betaflight gains and the best approximation using the SuperRate equation. With the current BetaFlight parameters all you can do is further flatten out the centre section but you are moving to no center stick response. I would need to see a response curve using both betaflight parameters that can't be approximated using the super rate curve.

**Comment by lthall** [comment]
> I wonder if we should take the opportunity to move the rate and expo params down into attitude control so we can share between copter and quadplane.

No, these are acro specific parameters. I would like to bring the ACRO parameters up to date to something like this:
ACRO_RP_EXPO - super rate equation
ACRO_Y_EXPO - super rate equation
ACRO_RP_RATE - degrees per second
ACRO_Y_RATE - degrees per second

Based on what I have seen so far I think this will take care of the pilots input shaping needs. I think anything more would just be providing more ways to achieve the same results and dramatically complicating the variables (and support costs).

For use in non-acro modes where we command yaw
PILOT_Y_RATE - degrees per second

I am wondering if we also need expo for non-ACRO modes, yaw rate and roll and pitch angle?

**Comment by chronowarper** [comment]
I would vote for having the same Yaw response in acro and non-acro mode, otherwise it would be strange to have the quad yaw differently based on which mode it was. I wonder if exponential on roll/pitch in stabilized mode would be nice to have.

Regarding the 3rd parameter: 
Without the BF expo I have a hard time getting this 'super-smooth' curve I ran on one of my quads. In particular it's hard to hit the 354 at 0.8 deflection with such a low sloped center. It's a really slow rate near the center so that's in line with what you observed: EXPO just flattens the curve. The way I was describing this wasn't very clear. 

x    - deg/s
0.2 - 26
0.4 - 66
0.6 - 155
0.8 - 354
1.0 - 772

Incidentally I found that the formula I copied from here (https://www.desmos.com/calculator/r5pkxlxhtb) for this code doesn't handle RC > 2.0 properly.

But in my BF configurator and this calculator (https://apocolipse.github.io/RotorPirates/), I can get this curve with RC = 2.1, Super = 0.08, Expo = 0.87... think I'll need to look at the BF code. 


For the other set of rates that I run for freestyle on rest of my quads, I am able to approximately get the same without EXPO.

**Comment by lthall** [comment]
This is the three curves with the same zero stick slope as your parameters:
image

The difference is how late they slope up to your maximum rate. Your use of only 0.08 means you are using the ^4 option effectively.

It sounds to me like you want earlier slope up rather than sick end slope up.

**Comment by lthall** [comment]
For a fixed high end slope this gives a good idea of the range that could be achieved with a two variable solution. It is about 10% maximum.
image

**Comment by chronowarper** [comment]
Sorry I may be using the wrong term in what I meant by "center-stick"... maybe I should have said "center and mid-stick". What I want is a later slope-up because I use up to 0.6~0.7 for most of my flying, and blip slightly more for an aggressive turn, and blip to the end for a half or full roll/flip. 

I think it's not as important to match the slope at the end stick (the last 10% of deflection); I would mostly be blipping the stick in that region. And matching the slope in center isn't enough since that's just a small range of operation. 

I think what is not apparent with the normalized rate curve to '1' is a 5-10% difference is huge when one uses very high end-stick rates (maybe I run rates higher than most people would run...) 10% of 1080deg/s is 108deg/s. 

image

Though this examples demonstrates curves can be equivalent, none of those are curves that I would use. I'm not sure how other miniquad pilots set their end rates up, but I would find these too aggressive center~mid-stick... even if I went back down to a end-stick of 780.

image

**Comment by lthall** [comment]
You said you use 
> RC = 2.1, Super = 0.08, Expo = 0.87.
image

compared to expo (x^4) only:
> RC = 2.1, Super = 0.00, Expo = 0.87.
image

compared to super rate only:
> RC = 2.1, Super = 0.87, Expo = 0.00.
image

the expo tends to have a continuous curve to the point you highlighted while the Super rate has a more linear curve.

**Comment by lthall** [comment]
Comparing Expo at 0.87 and super rate at the same output at 70% deflection:

image

Super rate more linear and more responsive about centre stick (zero input).

The lower order expo's can't reach this at all.

image

**Comment by chronowarper** [comment]
Ah sorry, you had asked for a counter example and that's all I gave, except for a footnote comment at the end of my post :) 

I've use these two rates so far this year: 
RC = 2.1, Super = 0.08, Expo = 0.87
RC = 2.08, Super = 0.66, Expo = 0.11

The first are the rates I used for constrained space or cine flying with some ability to pull sharp turns. I've started using the 2nd one for more open space acro. And as I get better at center~mid-stick control, I don't use the high expo as much.... so I'm on the same wavelength that Super is better in general. 

Your charts do help me understand the amount of value a third Expo param adds is limited and isn't as wide as I made it out to be. It may not be the most useful piece to add as-is. May be interesting to think about something else to adjust that 'transition' point if people find that meaningful.

**Comment by lthall** [comment]
> so I'm on the same wavelength that Super is better in general.

Cool. After busting this down I feel pretty confident that this will provide the most flexible solution while still providing a nice curve about centre stick (zero).

> Your charts do help me understand the amount of value a third Expo param adds is limited and isn't as wide as I made it out to be. It may not be the most useful piece to add as-is. May be interesting to think about something else to adjust that 'transition' point if people find that meaningful.

Yes the thing that is much harder to evaluate is the "value" and "cost" of a feature. We need to be able to clearly explain the feature and to be able to support it over the next ... well for ever. And we need to be able to explain to users the "value" of different combinations of settings. I can do that with a move to super expo equation. Using a two variable solution makes it much harder. I can't clearly explain how to combine the two variable to achieve a given result, why you want one vs another (in general people couldn't tell the difference in flight), much less explain to a third person how to answer those questions.

In short, I think I can justify and get super rate into the code now, along with directly specifying the angular rates directly instead of the stupid Rate P thing.

**Comment by lthall** [comment]
I am thinking something like this:
https://github.com/ArduPilot/ardupilot/pull/18689

**Comment by chronowarper** [comment]
Cool, thanks! The formula looks good. 

Regarding an earlier comment about expo in stabilized modes... I think if anyone wanted expo in roll/pitch in stabilized mode could just do it on the radio. Resolution isn't critical given it's at most it would map ~60 degrees to stick input.

**Comment by lthall** [comment]
@chronowarper Andy has test flown the code I suggested. I would be interested to hear if you can get your desired feel using just the parameters provided.

**Comment by chronowarper** [comment]
I built and flashed a binary from the pull (using STM32Programmer) and loaded my old params, did accelerometer calibration but I seem to be stuck at compass calibration. When I hit 'start', there's no ID that pops up. Not sure what I did wrong.

image

I set up a BF quad with just super rates and felt the rates were fairly good. I can confirm I can get the same rates with the Arducopter formula with excel (just showing one of the two rates) 

image

**Comment by lthall** [comment]
I am sorry but I don't know why you would see that.

**Comment by chronowarper** [comment]
I disabled the arming checks so I can fly. The rates feel good 👍

**Comment by lthall** [comment]
> I disabled the arming checks so I can fly. The rates feel good

Great! Thanks for your work on this!

**Comment by chronowarper** [comment]
> > I disabled the arming checks so I can fly. The rates feel good
> 
> Great! Thanks for your work on this!

Thank you and the dev team for reviewing and implementing the new expo rates!

**Comment by chronowarper** [comment]
Regarding the compass cal issue... I can't reproduce it. I tried doing some regression testing by flashing some older firmware and working back up to the PR, but can't reproduce with the same PR binary I was using yesterday.

**Comment by chronowarper** [comment]
I'm closing this pull request as PR #18689 implements super-rate in Arducopter.
